File: transformer_engine/pytorch/csrc/extensions/softmax.cpp
/*************************************************************************
 * Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 *
 * See LICENSE for license information.
 ************************************************************************/

#include <transformer_engine/softmax.h>

#include "../stable_common.h"

namespace transformer_engine::pytorch::stable {

using Tensor = torch::stable::Tensor;

Tensor scaled_softmax_forward(Tensor input, double scale_factor) {
  NVTE_CHECK(input.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(input, "scaled_softmax_forward");

  auto sizes = input.sizes();
  const int64_t batches = sizes[0];
  const int64_t attn_heads = sizes[1];
  const int64_t query_seq_len = sizes[2];
  const int64_t key_seq_len = sizes[3];

  NVTE_CHECK(key_seq_len <= 16384, "Key sequence length must be 16384 or less");
  NVTE_CHECK(key_seq_len % 8 == 0, "Key sequence length must be divisible by 8");
  NVTE_CHECK(query_seq_len > 1, "Query sequence length must be greater than 1");

  auto softmax_results = allocateStableTensor({batches, attn_heads, query_seq_len, key_seq_len},
                                              input.scalar_type(), input.get_device_index());

  auto input_cu = makeTransformerEngineTensor(input);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_softmax_forward(input_cu.data(), softmax_results_cu.data(),
                              static_cast<float>(scale_factor),
                              getCurrentCUDAStreamRaw(input.get_device_index()));

  return softmax_results;
}

Tensor scaled_softmax_backward(Tensor output_grad, Tensor softmax_results, double scale_factor) {
  auto output_grads = torch::stable::contiguous(output_grad);
  softmax_results = torch::stable::contiguous(softmax_results);

  NVTE_CHECK(output_grads.dim() == 4, "expected 4D tensor");
  NVTE_CHECK(softmax_results.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(output_grads, "scaled_softmax_backward");
  check_fp16_bf16(softmax_results, "scaled_softmax_backward");

  auto output_grads_cu = makeTransformerEngineTensor(output_grads);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  // Produce gradients in place.
  nvte_scaled_softmax_backward(output_grads_cu.data(), softmax_results_cu.data(),
                               output_grads_cu.data(), static_cast<float>(scale_factor),
                               getCurrentCUDAStreamRaw(output_grads.get_device_index()));

  return output_grads;
}

Tensor scaled_masked_softmax_forward(Tensor input, Tensor mask, double scale_factor) {
  NVTE_CHECK(input.dim() == 4, "expected 4D tensor");
  NVTE_CHECK(mask.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(input, "scaled_masked_softmax_forward");
  input = torch::stable::contiguous(input);
  mask = torch::stable::contiguous(mask);

  auto sizes = input.sizes();
  const int64_t batches = sizes[0];
  const int64_t attn_heads = sizes[1];
  const int64_t query_seq_len = sizes[2];
  const int64_t key_seq_len = sizes[3];

  auto mask_sizes = mask.sizes();
  const int64_t pad_batches = mask_sizes[0];
  NVTE_CHECK(pad_batches == 1 || pad_batches == batches,
             "Mask batch dim must be 1 or match input batch dim");
  NVTE_CHECK(mask_sizes[1] == 1, "Mask second dim must be 1");
  NVTE_CHECK(mask_sizes[2] == query_seq_len, "Mask query dim must match input");
  NVTE_CHECK(mask_sizes[3] == key_seq_len, "Mask key dim must match input");

  NVTE_CHECK(key_seq_len <= 16384, "Key sequence length must be 16384 or less");
  NVTE_CHECK(key_seq_len % 8 == 0, "Key sequence length must be divisible by 8");
  NVTE_CHECK(query_seq_len > 1, "Query sequence length must be greater than 1");

  auto softmax_results = allocateStableTensor({batches, attn_heads, query_seq_len, key_seq_len},
                                              input.scalar_type(), input.get_device_index());

  auto input_cu = makeTransformerEngineTensor(input);
  auto mask_cu = makeTransformerEngineTensor(mask);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_masked_softmax_forward(input_cu.data(), mask_cu.data(), softmax_results_cu.data(),
                                     static_cast<float>(scale_factor),
                                     getCurrentCUDAStreamRaw(input.get_device_index()));

  return softmax_results;
}

Tensor scaled_masked_softmax_backward(Tensor output_grad, Tensor softmax_results,
                                      double scale_factor) {
  auto output_grads = torch::stable::contiguous(output_grad);
  softmax_results = torch::stable::contiguous(softmax_results);

  NVTE_CHECK(output_grads.dim() == 4, "expected 4D tensor");
  NVTE_CHECK(softmax_results.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(output_grads, "scaled_masked_softmax_backward");
  check_fp16_bf16(softmax_results, "scaled_masked_softmax_backward");

  auto output_grads_cu = makeTransformerEngineTensor(output_grads);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_softmax_backward(output_grads_cu.data(), softmax_results_cu.data(),
                               output_grads_cu.data(), static_cast<float>(scale_factor),
                               getCurrentCUDAStreamRaw(output_grads.get_device_index()));

  return output_grads;
}

Tensor scaled_upper_triang_masked_softmax_forward(Tensor input, double scale_factor) {
  NVTE_CHECK(input.dim() == 3, "expected 3D tensor");
  check_fp16_bf16(input, "scaled_upper_triang_masked_softmax_forward");

  auto sizes = input.sizes();
  const int64_t attn_batches = sizes[0];
  const int64_t seq_len = sizes[1];
  NVTE_CHECK(seq_len <= 16384, "Sequence length must be 16384 or less");

  auto softmax_results = allocateStableTensor({attn_batches, seq_len, seq_len}, input.scalar_type(),
                                              input.get_device_index());

  auto input_cu = makeTransformerEngineTensor(input);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_upper_triang_masked_softmax_forward(
      input_cu.data(), softmax_results_cu.data(), static_cast<float>(scale_factor),
      getCurrentCUDAStreamRaw(input.get_device_index()));

  return softmax_results;
}

Tensor scaled_upper_triang_masked_softmax_backward(Tensor output_grads_, Tensor softmax_results_,
                                                   double scale_factor) {
  auto output_grads = torch::stable::contiguous(output_grads_);
  auto softmax_results = torch::stable::contiguous(softmax_results_);

  NVTE_CHECK(output_grads.dim() == 3, "expected 3D tensor");
  NVTE_CHECK(softmax_results.dim() == 3, "expected 3D tensor");
  check_fp16_bf16(output_grads, "scaled_upper_triang_masked_softmax_backward");
  check_fp16_bf16(softmax_results, "scaled_upper_triang_masked_softmax_backward");
  NVTE_CHECK(output_grads.sizes()[1] == output_grads.sizes()[2],
             "Upper triangular softmax requires square attention matrix");

  auto output_grads_cu = makeTransformerEngineTensor(output_grads);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_upper_triang_masked_softmax_backward(
      output_grads_cu.data(), softmax_results_cu.data(), output_grads_cu.data(),
      static_cast<float>(scale_factor), getCurrentCUDAStreamRaw(output_grads.get_device_index()));

  return output_grads;
}

Tensor scaled_aligned_causal_masked_softmax_forward(Tensor input, double scale_factor) {
  NVTE_CHECK(input.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(input, "scaled_aligned_causal_masked_softmax_forward");

  auto sizes = input.sizes();
  const int64_t batches = sizes[0];
  const int64_t attn_heads = sizes[1];
  const int64_t query_seq_len = sizes[2];
  const int64_t key_seq_len = sizes[3];

  NVTE_CHECK(key_seq_len <= 16384, "Key sequence length must be 16384 or less");
  NVTE_CHECK(key_seq_len % 8 == 0, "Key sequence length must be divisible by 8");
  NVTE_CHECK(query_seq_len >= 1, "Query sequence length must be greater or equal to 1");

  auto softmax_results = allocateStableTensor({batches, attn_heads, query_seq_len, key_seq_len},
                                              input.scalar_type(), input.get_device_index());

  auto input_cu = makeTransformerEngineTensor(input);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_aligned_causal_masked_softmax_forward(
      input_cu.data(), softmax_results_cu.data(), static_cast<float>(scale_factor),
      getCurrentCUDAStreamRaw(input.get_device_index()));

  return softmax_results;
}

Tensor scaled_aligned_causal_masked_softmax_backward(Tensor output_grad, Tensor softmax_results_,
                                                     double scale_factor) {
  auto output_grads = torch::stable::contiguous(output_grad);
  auto softmax_results = torch::stable::contiguous(softmax_results_);

  NVTE_CHECK(output_grads.dim() == 4, "expected 4D tensor");
  NVTE_CHECK(softmax_results.dim() == 4, "expected 4D tensor");
  check_fp16_bf16(output_grads, "scaled_aligned_causal_masked_softmax_backward");
  check_fp16_bf16(softmax_results, "scaled_aligned_causal_masked_softmax_backward");

  auto output_grads_cu = makeTransformerEngineTensor(output_grads);
  auto softmax_results_cu = makeTransformerEngineTensor(softmax_results);

  nvte_scaled_aligned_causal_masked_softmax_backward(
      output_grads_cu.data(), softmax_results_cu.data(), output_grads_cu.data(),
      static_cast<float>(scale_factor), getCurrentCUDAStreamRaw(output_grads.get_device_index()));

  return output_grads;
}

}  // namespace transformer_engine::pytorch::stable

STABLE_TORCH_LIBRARY_IMPL(transformer_engine, CUDA, m) {
  m.impl("scaled_softmax_forward",
         TORCH_BOX(&transformer_engine::pytorch::stable::scaled_softmax_forward));
  m.impl("scaled_softmax_backward",
         TORCH_BOX(&transformer_engine::pytorch::stable::scaled_softmax_backward));
  m.impl("scaled_masked_softmax_forward",
         TORCH_BOX(&transformer_engine::pytorch::stable::scaled_masked_softmax_forward));
  m.impl("scaled_masked_softmax_backward",
         TORCH_BOX(&transformer_engine::pytorch::stable::scaled_masked_softmax_backward));
  m.impl(
      "scaled_upper_triang_masked_softmax_forward",
      TORCH_BOX(&transformer_engine::pytorch::stable::scaled_upper_triang_masked_softmax_forward));
  m.impl(
      "scaled_upper_triang_masked_softmax_backward",
      TORCH_BOX(&transformer_engine::pytorch::stable::scaled_upper_triang_masked_softmax_backward));
  m.impl("scaled_aligned_causal_masked_softmax_forward",
         TORCH_BOX(
             &transformer_engine::pytorch::stable::scaled_aligned_causal_masked_softmax_forward));
  m.impl("scaled_aligned_causal_masked_softmax_backward",
         TORCH_BOX(
             &transformer_engine::pytorch::stable::scaled_aligned_causal_masked_softmax_backward));
}

File: transformer_engine/pytorch/attention/dot_product_attention/softmax.py
# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# See LICENSE for license information.

"""Fused scaled masked softmax functions"""
import os
from typing import Callable, Optional
import torch
from torch import nn
import transformer_engine_torch as tex
from transformer_engine.pytorch.export import is_in_onnx_export_mode


THREADS_PER_WARP = 32
THREADS_PER_BLOCK = 128


_default_causal_mask = {}


def _scale_to_tensor(scale: float) -> torch.Tensor:
    """Wrap a Python float in a 0-D tensor expected by the tex kernels."""
    return torch.tensor([scale])[0]


# ----------------------------- ScaledSoftmax -------------------------------


@torch.library.custom_op("te_softmax::scaled_softmax_fwd", mutates_args=())
def scaled_softmax_forward(inputs: torch.Tensor, scale: float) -> torch.Tensor:
    """Forward pass for ScaledSoftmax."""
    return tex.scaled_softmax_forward(inputs, _scale_to_tensor(scale))


@scaled_softmax_forward.register_fake
def _scaled_softmax_forward_fake(inputs: torch.Tensor, scale: float) -> torch.Tensor:
    del scale
    return torch.empty_like(inputs)


@torch.library.custom_op("te_softmax::scaled_softmax_bwd", mutates_args=())
def scaled_softmax_backward(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    """Backward pass for ScaledSoftmax."""
    return tex.scaled_softmax_backward(output_grads, softmax_results, _scale_to_tensor(scale))


@scaled_softmax_backward.register_fake
def _scaled_softmax_backward_fake(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    del softmax_results, scale
    return torch.empty_like(output_grads)


def _scaled_softmax_setup_context(ctx, inputs, output):
    _inp, scale = inputs
    ctx.scale = scale
    ctx.save_for_backward(output)


def _scaled_softmax_backward_wrapper(ctx, grad_output):
    (softmax_results,) = ctx.saved_tensors
    grad_inputs = torch.ops.te_softmax.scaled_softmax_bwd(grad_output, softmax_results, ctx.scale)
    return grad_inputs, None


scaled_softmax_forward.register_autograd(
    _scaled_softmax_backward_wrapper,
    setup_context=_scaled_softmax_setup_context,
)


# --------------------------- ScaledMaskedSoftmax ---------------------------


@torch.library.custom_op("te_softmax::scaled_masked_softmax_fwd", mutates_args=())
def scaled_masked_softmax_forward(
    inputs: torch.Tensor, mask: torch.Tensor, scale: float
) -> torch.Tensor:
    """Forward pass for ScaledMaskedSoftmax."""
    return tex.scaled_masked_softmax_forward(inputs, mask, _scale_to_tensor(scale))


@scaled_masked_softmax_forward.register_fake
def _scaled_masked_softmax_forward_fake(
    inputs: torch.Tensor, mask: torch.Tensor, scale: float
) -> torch.Tensor:
    del mask, scale
    return torch.empty_like(inputs)


@torch.library.custom_op("te_softmax::scaled_masked_softmax_bwd", mutates_args=())
def scaled_masked_softmax_backward(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    """Backward pass for ScaledMaskedSoftmax."""
    return tex.scaled_masked_softmax_backward(
        output_grads, softmax_results, _scale_to_tensor(scale)
    )


@scaled_masked_softmax_backward.register_fake
def _scaled_masked_softmax_backward_fake(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    del softmax_results, scale
    return torch.empty_like(output_grads)


def _scaled_masked_softmax_setup_context(ctx, inputs, output):
    _inp, _mask, scale = inputs
    ctx.scale = scale
    ctx.save_for_backward(output)


def _scaled_masked_softmax_backward_wrapper(ctx, grad_output):
    (softmax_results,) = ctx.saved_tensors
    grad_inputs = torch.ops.te_softmax.scaled_masked_softmax_bwd(
        grad_output, softmax_results, ctx.scale
    )
    return grad_inputs, None, None


scaled_masked_softmax_forward.register_autograd(
    _scaled_masked_softmax_backward_wrapper,
    setup_context=_scaled_masked_softmax_setup_context,
)


# ---------------------- ScaledUpperTriangMaskedSoftmax ----------------------


@torch.library.custom_op("te_softmax::scaled_upper_triang_masked_softmax_fwd", mutates_args=())
def scaled_upper_triang_masked_softmax_forward(
    inputs: torch.Tensor, scale: float
) -> torch.Tensor:
    """Forward pass for ScaledUpperTriangMaskedSoftmax."""
    return tex.scaled_upper_triang_masked_softmax_forward(inputs, _scale_to_tensor(scale))


@scaled_upper_triang_masked_softmax_forward.register_fake
def _scaled_upper_triang_masked_softmax_forward_fake(
    inputs: torch.Tensor, scale: float
) -> torch.Tensor:
    del scale
    return torch.empty_like(inputs)


@torch.library.custom_op("te_softmax::scaled_upper_triang_masked_softmax_bwd", mutates_args=())
def scaled_upper_triang_masked_softmax_backward(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    """Backward pass for ScaledUpperTriangMaskedSoftmax."""
    return tex.scaled_upper_triang_masked_softmax_backward(
        output_grads, softmax_results, _scale_to_tensor(scale)
    )


@scaled_upper_triang_masked_softmax_backward.register_fake
def _scaled_upper_triang_masked_softmax_backward_fake(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    del softmax_results, scale
    return torch.empty_like(output_grads)


def _scaled_upper_triang_masked_softmax_setup_context(ctx, inputs, output):
    _inp, scale = inputs
    ctx.scale = scale
    ctx.save_for_backward(output)


def _scaled_upper_triang_masked_softmax_backward_wrapper(ctx, grad_output):
    (softmax_results,) = ctx.saved_tensors
    grad_inputs = torch.ops.te_softmax.scaled_upper_triang_masked_softmax_bwd(
        grad_output, softmax_results, ctx.scale
    )
    return grad_inputs, None


scaled_upper_triang_masked_softmax_forward.register_autograd(
    _scaled_upper_triang_masked_softmax_backward_wrapper,
    setup_context=_scaled_upper_triang_masked_softmax_setup_context,
)


# -------------------- ScaledAlignedCausalMaskedSoftmax ---------------------


@torch.library.custom_op("te_softmax::scaled_aligned_causal_masked_softmax_fwd", mutates_args=())
def scaled_aligned_causal_masked_softmax_forward(
    inputs: torch.Tensor, scale: float
) -> torch.Tensor:
    """Forward pass for ScaledAlignedCausalMaskedSoftmax."""
    return tex.scaled_aligned_causal_masked_softmax_forward(inputs, _scale_to_tensor(scale))


@scaled_aligned_causal_masked_softmax_forward.register_fake
def _scaled_aligned_causal_masked_softmax_forward_fake(
    inputs: torch.Tensor, scale: float
) -> torch.Tensor:
    del scale
    return torch.empty_like(inputs)


@torch.library.custom_op("te_softmax::scaled_aligned_causal_masked_softmax_bwd", mutates_args=())
def scaled_aligned_causal_masked_softmax_backward(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    """Backward pass for ScaledAlignedCausalMaskedSoftmax."""
    return tex.scaled_aligned_causal_masked_softmax_backward(
        output_grads, softmax_results, _scale_to_tensor(scale)
    )


@scaled_aligned_causal_masked_softmax_backward.register_fake
def _scaled_aligned_causal_masked_softmax_backward_fake(
    output_grads: torch.Tensor, softmax_results: torch.Tensor, scale: float
) -> torch.Tensor:
    del softmax_results, scale
    return torch.empty_like(output_grads)


def _scaled_aligned_causal_masked_softmax_setup_context(ctx, inputs, output):
    _inp, scale = inputs
    ctx.scale = scale
    ctx.save_for_backward(output)


def _scaled_aligned_causal_masked_softmax_backward_wrapper(ctx, grad_output):
    (softmax_results,) = ctx.saved_tensors
    grad_inputs = torch.ops.te_softmax.scaled_aligned_causal_masked_softmax_bwd(
        grad_output, softmax_results, ctx.scale
    )
    return grad_inputs, None


scaled_aligned_causal_masked_softmax_forward.register_autograd(
    _scaled_aligned_causal_masked_softmax_backward_wrapper,
    setup_context=_scaled_aligned_causal_masked_softmax_setup_context,
)


_default_causal_mask = {}


def _get_default_causal_mask(mask_type: str, sq: int, sk: int) -> torch.Tensor:
    """Return the causal upper triangular mask for softmax input"""

    def _get_mask():
        diagonal_offset = sk - sq + 1 if "bottom_right" in mask_type else 1
        return torch.triu(
            torch.ones(sq, sk, dtype=torch.bool, device="cuda"), diagonal=diagonal_offset
        )

    if is_in_onnx_export_mode():
        return _get_mask()
    matrix_identifiers = (mask_type, sq, sk)
    if matrix_identifiers not in _default_causal_mask:
        _default_causal_mask[matrix_identifiers] = _get_mask()
    return _default_causal_mask[matrix_identifiers]


class FusedScaleMaskSoftmax(nn.Module):
    """
    fused operation: scaling + mask + softmax

    Arguments:
        mask_func: mask function to be applied.
        softmax_in_fp32: if true, softmax in performed at fp32 precision.
    """

    def __init__(
        self,
        mask_func: Callable,
        softmax_in_fp32: bool = True,
    ) -> None:
        super().__init__()
        self.scaled_masked_softmax_fusion_type = bool(
            int(os.getenv("NVTE_MASKED_SOFTMAX_FUSION", "1"))
        )
        self.mask_func = mask_func
        self.softmax_in_fp32 = softmax_in_fp32

    def forward(
        self,
        inp: torch.Tensor,
        mask: torch.Tensor,
        attn_mask_type: str,
        scale: Optional[float] = None,
    ) -> torch.Tensor:
        """FusedScaleMaskSoftmax fprop"""
        # [b, np, sq, sk]
        assert inp.dim() == 4
        self.input_in_fp16 = inp.dtype == torch.float16
        self.input_in_bf16 = inp.dtype == torch.bfloat16
        self.input_in_float16 = self.input_in_fp16 or self.input_in_bf16
        self.attn_mask_type = attn_mask_type

        assert scale is None or self.softmax_in_fp32, "softmax should be in fp32 when scaled"
        if is_in_onnx_export_mode():
            return self.forward_torch_softmax(inp, mask, scale)

        # We do not want to connect this if with previous if,
        # because we want to avoid calling is_kernel_available() in ONNX mode.
        if self.is_kernel_available(mask, *inp.size()):
            return self.forward_fused_softmax(inp, mask, scale)
        return self.forward_torch_softmax(inp, mask, scale)

    def is_kernel_available(self, mask: torch.Tensor, b: int, np: int, sq: int, sk: int) -> bool:
        """Check FusedScaleMaskSoftmax kernel availability based on size"""
        attn_batches = b * np

        if not self.scaled_masked_softmax_fusion_type:
            return False  # user doesn't want to fuse
        if not self.input_in_float16:
            return False  # input must be fp16
        if not 16 < sk < 16384:
            return False  # sk must be 16 ~ 16384
        if sk % 8 != 0:
            return False  # sk must be divisor of 8
        if sq == 1:
            return False  # sq must be > 1
        if self.attn_mask_type == "causal" and sq != sk:
            return False  # Fused causal kernel only support causal_bottom_right

        if (
            sq % 4 == 0  # sq must be divisor of 4
            and attn_batches % 4 == 0  # np * b must be divisor of 4
        ):
            batch_per_block = self.get_batch_per_block(int(sk))
            if "padding" in self.attn_mask_type or self.attn_mask_type == "arbitrary":
                if (
                    mask is not None
                    and sq % batch_per_block == 0
                    and mask.shape[0] in [1, b]
                    and mask.shape[1:] == (1, sq, sk)
                ):
                    return True
            else:
                if sq % batch_per_block == 0:
                    return True
        return False

    def forward_fused_softmax(
        self, inp: torch.Tensor, mask: torch.Tensor, scale: Optional[float] = None
    ) -> torch.Tensor:
        """
        Fused masked softmax path.
          attn_mask_type                                       | module
        -----------------------------------------------------------------------------------------
          no_mask                                              | ScaledSoftmax
          causal (self-attention), causal_bottom_right         | ScaledAlignedCausalMaskedSoftmax
          padding, padding_causal, padding_causal_bottom_right | ScaledMaskedSoftmax
          arbitrary ([1, 1, sq, sk] or [b, 1, sq, sk])         | ScaledMaskedSoftmax
        """
        scale = 1.0 if scale is None else float(scale)

        # Disable for now until unalignment bug is fixed.
        # if self.attn_mask_type in ["causal", "causal_bottom_right"]:
        #    return torch.ops.te_softmax.scaled_aligned_causal_masked_softmax_fwd(inp, scale)

        # input is 4D tensor (1, 1, sq, sk) or (b, 1, sq, sk)
        if mask is not None and self.attn_mask_type != "no_mask":
            return torch.ops.te_softmax.scaled_masked_softmax_fwd(inp, mask, scale)
        return torch.ops.te_softmax.scaled_softmax_fwd(inp, scale)

    def forward_torch_softmax(
        self, inp: torch.Tensor, mask: torch.Tensor, scale: Optional[float] = None
    ) -> torch.Tensor:
        """Framework softmax"""
        if self.input_in_float16 and self.softmax_in_fp32:
            inp = inp.float()

        if scale is not None:
            inp = inp * scale

        if self.attn_mask_type in ["causal", "causal_bottom_right"]:
            seq_len_q, seq_len_k = inp.size(2), inp.size(3)
            causal_mask = _get_default_causal_mask(self.attn_mask_type, seq_len_q, seq_len_k)

            if mask is None:
                mask = causal_mask
            else:
                mask = torch.logical_or(mask, causal_mask)
        mask_output = inp
        if mask is not None and self.attn_mask_type != "no_mask":
            mask_output = self.mask_func(inp, mask)
        probs = torch.nn.functional.softmax(mask_output, dim=-1)

        if self.input_in_float16 and self.softmax_in_fp32:
            if self.input_in_fp16:
                probs = probs.half()
            else:
                probs = probs.bfloat16()

        return probs

    @staticmethod
    def get_batch_per_block(key_seq_len: int) -> int:
        """Softmax utility"""
        pow2 = 1 << (key_seq_len - 1).bit_length()
        warp_size = pow2 if pow2 < THREADS_PER_WARP else THREADS_PER_WARP
        batches_per_warp = 2 if pow2 <= 128 else 1
        warps_per_block = THREADS_PER_BLOCK // warp_size
        batches_per_block = warps_per_block * batches_per_warp
        return batches_per_block
